inc/Engine/CriticalPath/ProcessorService.php

```diff
@@ -40,24 +40,31 @@ public function __construct( DataManager $data_manager, APIClient $api_client )
 	 *
 	 * @param string $item_url  URL for item to be used in error messages.
 	 * @param string $item_path Path for item to be processed.
-	 * @param bool   $timeout   Timeout is requested or not.
-	 * @param bool   $is_mobile If this request is for mobile cpcss.
+	 * @param array  $additional_parameters additional parameters for generation.
 	 *
 	 * @return array|WP_Error
 	 */
-	public function process_generate( $item_url, $item_path, $timeout = false, $is_mobile = false ) {
+	public function process_generate( $item_url, $item_path, $additional_parameters = [
+		'timeout'   => false,
+		'is_mobile' => false,
+		'item_type' => 'custom',
+	] ) {
+		$timeout   = isset( $additional_parameters['timeout'] ) ? $additional_parameters['timeout'] : false;
+		$is_mobile = isset( $additional_parameters['is_mobile'] ) ? $additional_parameters['is_mobile'] : false;
+		$item_type = isset( $additional_parameters['item_type'] ) ? $additional_parameters['item_type'] : 'custom';
```

Comment:
Instead of doing it like this, we could use a `$defaults` array with the default value, and then use [wp_parse_args()](https://developer.wordpress.org/reference/functions/wp_parse_args/) to merge the parameter array with the default array:

`$args = wp_parse_args( $additional_parameters, $defaults );`

That way we always know that the index are all created with a value. What do you think?

Reply:
Amazing, totally right.
Adjusting it right now

Reply:
Or an even faster way is to use `array_merge()`.  Why? You know you have an array that needs to be merged with a default array. `wp_parse_args()` is nice when merging different data types. But when it's an array, it uses `array_merge()`. For us, we can skip all the checks in `wp_parse_args()` and just use PHP's `array_merge()` directly. That will be faster and use a little less memory.

Reply:
Yes I opened the implementation for `wp_parse_args` and it converts the first argument to array based on its type then apply `array_merge` on this argument array and defaults (second argument).
and mainly at unit testing I should mock it by using array_merge.
I imagined that using wordpress abstraction layer is good enough but I think on this case both are the same (from abstraction prespective).
Action: I will use `array_merge`

Reply:
Agreed